lib/open-search-client.ts

```diff
@@ -0,0 +1,378 @@
+ * document index fields. This query takes document key/values and finds
+ * the terms that match those field values. It returns a modified
+ * percolator query result.
+ * See https://www.elastic.co/guide/en/elasticsearch/reference/6.8/query-dsl-percolate-query.html
+ * for more information.
+ *
+ * @param prefixUrl - Open Search endpoint.
+ * @param fields - Key/value index fields to use to match
+ * the percolator query.
+ * @param options - Options for the request.
+ *
+ * @returns Modified percolator query result.
+ * The result will be a list of objects that includes the matched term
+ * label, uri, and source_terminology.
+ */
+export const percolateDocumentFields = async (
+  prefixUrl: string,
+  fields: { title: string, contents: string },
+  options: PercolateDocumentFieldsOptions = {}
+): Promise<DocumentItemTerm[]> => {
+  const {
+    from = 0,
+    size = 300,
+  } = options;
+
+  const body = {
+    query: {
+      percolate: {
+        field: 'query',
+        document: {
+          ...fields,
+        },
+      },
+    },
+    from,
+    size,
+  };
+
+  const rawPercolateResponse = await got4aws().post(
+    'terms/_search',
+    {
+      prefixUrl,
+      json: body,
+      responseType: 'json',
+      resolveBodyOnly: true,
+    }
+  );
+
+  const percolateResponse = percolateResponseSchema
+    .safeParse(rawPercolateResponse);
+
+  if (!percolateResponse.success) {
+    console.log(`ERROR: Failed to parse percolate response: ${percolateResponse.error}`);
+    throw percolateResponse.error;
+  }
+
+  // It's possible we'll tweak this response as we improve the
+  // percolator query. Leaving this as is for now.
+  const { hits: { hits } } = percolateResponse.data;
+  return hits.map((h) => ({
+    label: h._source.query.multi_match.query,
+    uri: h._id,
+    source_terminology: h._source.source_terminology,
+  }));
+};
+
+/**
+ * @param prefixUrl
+ * @param index
+ * @returns
+ */
+export const getIndex = async (
+  prefixUrl: string,
+  index: string
+): Promise<unknown> => gotEs(prefixUrl).get(index);
+
+/**
+  * @param prefixUrl
+  * @param index
+  * @param indexBody
+  * @returns
+  */
+export const createIndex = (
+  prefixUrl: string,
+  index: string,
+  indexBody?: Record<string, unknown>
+): Promise<unknown> =>
+  gotEs(prefixUrl).put(
+    index,
+    {
+      json: indexBody,
+      resolveBodyOnly: false,
+      throwHttpErrors: false,
+    }
+  ).then(({ statusCode, body }) => {
+    if (statusCode === 200) return body;
+
+    const errorMessage = get(
+      body,
+      'error.type',
+      `Unexpected ${statusCode} response: ${body}`
+    );
+
+    throw new Error(errorMessage);
+  });
+
+/**
+ * Indexes an index item into the documents index.
+ *
+ * @param prefixUrl - Open Search endpoint.
+ * @param documentItem - Object to index.
+ * @returns
+ */
+export const putDocumentItem = async (
+  prefixUrl: string,
+  documentItem: DocumentItem
+): Promise<PutDocumentItemResponse> => got4aws().post(
```

Comment:
We should be validating this response type.

Reply:
I've just marked these as `Promise<unknown>` for now. We're not using this response anywhere right now and if we do I can improve that validation and return type.

Reply:
I take that back. I improved the put document response schema and correctly validate and return it.